builder.md

```diff
@@ -1,57 +1,188 @@
+executed, and MUST look like `<output-dir>/base16-<slug><extension>`, where the
+`SLUG` is taken from the scheme filename made lowercase with spaces replaced
+with dashes and both the extension and `output-dir` are taken from
+`config.yaml`.
+
+Build failures are mostly related to faulty schemes and/or templates:
+- when failling to parse a scheme, the program MUST exit with code `1`;
+- when failling to parse a template file, the program MUST exit with code `2`;
+- otherwise, the program MUST exit with code `3`.
+
+All of these errors MAY be accompanied by OPTIONAL error messages.
+
+The exit codes are not exhaustive and new ones might be added in the future, so
+scripters catching generic errors SHOULD check for non-zero status.
+
+Otherwise, the program MUST exit with code `0` and an OPTIONAL success message.
+
+#### 3.1.3 CLI example:
+
+```bash
+git clone https://github.com/base16-project/base16-vim.git
+git clone https://github.com/base16-project/base16-schemes.git
+base16 base16-vim/templates base16-schemes/*.yaml
+```
+
+Should create a `colors` (on the current directory) containing `base16-XXX.vim`
+files.
+
+### 3.2. Library
+
+The other option is exposing a software library other developers may use to
+assist developing more complex base16-compatible tooling.
+
+As above, the library MUST implement single feature: building templates.
+
+This exposed library, or any internal code, has no specific required structure.
+
+The author MAY choose how (through a class, function that operates on file
+paths, etc) they will expose the builder functionality to the caller, according
+to their judgement.
+
+They SHOULD try to achieve an ergonomical and powerful interface and follow the
+best practices on their respective programming languages.
 
-## Code Structure
-There is no outline for a recommended code structure that a base16 theme builder should follow but a design goal should be to keep the application as simple as possible providing only the functionality described in this document. If you feel you have a great idea for additional functionality please raise an issue in the [base16 repository](https://github.com/chriskempson/base16).
+Official reference implementations will follow semantic versioning matching
+this spec, in the format: `<base16-version>-builder_revision`.
 
-## Considerations
-Mustache was chosen as the templating language due to its simplicity and widespread adoption across languages. YAML was chosen to describe scheme and configuration files for the same reasons.
+## 4. Considerations
+Mustache was chosen as the templating language due to its simplicity and
+widespread adoption across languages. YAML was chosen to describe scheme and
+configuration files for the same reasons.
 
-The core scheme repository was based off the single scheme repository so builders supporting v0.8-v0.9 of the spec can continue to function without changes.
+Spec upgrades should be backwards compatible with existing templates and
```

Comment:
I'm becoming less and less enamored with a builder spec... esp at the level actionless is desiring (and what perhaps we had before)... I think our goal (over time) should be to preserve compatibility with converting schemes into themes using templates... not that the exact *command line arguments* for doing so should be set in stone.

I find this an onerous requirement and I honestly don't feel like the project or it's contributors have that kind of time.

I think if someone wants to maintain such a thing they should do it with some small wrapper scripts around one of the official builders... I don't think it needs to be baked into the Base16 spec itself.

Reply:
> I'm becoming less and less enamored with a builder spec... esp at the level actionless is desiring (and what perhaps we had before)... I think our goal (over time) should be to preserve compatibility with converting schemes into themes using templates generally speaking... not that the exact command line arguments and process for doing so should be set in stone.

It's my personal opinion that the spec should be limited to purely the template format and the variables which should be provided. Having the config.yaml and how to handle output is helpful, and might make sense to keep around, but isn't the primary purpose of this spec.

> I find this an onerous requirement and I honestly don't feel like the project or it's contributors have that kind of time.

I would agree with this - it's a very large requirement. Additionally, as someone who maintains a builder, I like adding new features and supporting things. Having it this limited removes some of the fun for me.

> I think if someone wants to maintain such a thing they should do it with some small wrapper scripts around one of the official builders... I don't think it needs to be baked into the Base16 spec itself.

I think this makes sense. Maybe we could clarify that this spec is focused on a how to build a template rather than provide a complete interface to a CLI builder.

Reply:
i still think it should be return to the spec but just under new title, "Template management [Optional]" - it never was causing any issues in original base16 builders even when it was required - so i can't imagine how it would hurt when it's optional

Reply:
Everything not in the spec is already optional, by default. 😎 If you want it, built it. 🛠

Reply:
oh wow you're so funny guy 😹😹😹

Reply:
> > I find this an onerous requirement and I honestly don't feel like the project or it's contributors have that kind of time.
> 
> I would agree with this - it's a very large requirement. Additionally, as someone who maintains a builder, I like adding new features and supporting things. Having it this limited removes some of the fun for me.

I would argue that my proposal is for exactlying making this kind of feature additions easy!

By decoupling _what implements the spec_ and _satellite convenience features_, you could easily add more (even backwards-incompatible) features to your builder to cater, while still offering a stable and easy to script option! In this case you would split your current builder into a _reference builder_  and a higher level tool (manager? extended builder?).

Reply:
you also not following the main problem here - it was before fully documented in spec, but now you offering to just base the implementation on guesses

which will lead what different builders eventually will handle that part differently for no reason

> exactlying making this kind of feature additions easy!

we are not talking about ADDITIONS of features, but about returning back what was accidentially removed from spec while it was almost no maintainers/community around to argue that weird decision

Reply:
My main change was specifying what would be a _reference builder_ VS a broader base16 tool.

This is not backwards-incompatible. Old builders will still build. This change refers _specifically_ to the implementation of _new_, _reference_ builders.

Reply:
I proposed new reference builders be simple and unixy. That is, they shouldn't include the _manager_ previous "builders" (quoted as they did more than just build) included.

Reply:
Even if this was agreed upon, builders with more feature would still be builders and would still be compliant.

Reply:
I don't think file management should be in the specification. Builders are free to include them, of course, but it makes no sense to standartize something that is targeted at opinionated usage.

Reply:
> accidentially removed f

It also wasn't accidentally removed.  It was _removed with intention_, but the fact that a very few people actually use the "build world" use case wasn't considered... so now the question is what to do now... just because someone was once in the spec doesn't mean it always should be.

> Old builders will still build. 

That's a good point, if @actionless wants the old behavior they can use an older builder (for now)...  but I'm in total agreement with you that "build world" should be built on smaller pieces... and it may not even be our responsibility to build/spec it if it's only needed by a few people...

Reply:
> I don't think file management should be in the specification. Builders are free to include them, of course, but it makes no sense to standardize something that is targeted at opinionated usage.

> It was removed with intention

and why you think you have a better overview on this question than people who previously implemented that feature in original base16?

the reason why base16 project decayed because original maintainer was the only owner, not because it was numerous complains of this things en-complicating the spec

your arguments and motivation are like if i proposing a new feature, while i'm proposing to **fix** thing which was working for years and just got removed for no reason - deliberaly breaking backward-compatibility requirement of a spec

you failing to explain any real reasons to remove it - only what you dont understand why other people are using it, while you are not using it